### PR TITLE
Modify embedding to accept arbitrary input dim

### PR DESCRIPTION
**Pull request by linxihui**
1. Support high dimension input, e.g., in NLP, input could have shape like `(n_batch, n_sentences, n_words)`. 
2. No need to use `input_length` at all, but left for back compatibility.

### CODE REVIEW

**Comment by fchollet** [comment]
We absolutely do need support for `input_length` (used when the input length is static).

**Comment by linxihui** [comment]
@fchollet  Can you give an example that `input_length` is a must in this embedding layer? For something like connecting to a `Flatten` followed by a `Dense` layer? 

In my opinion, whether the `input_length` should be static or not should depend on if the `input_shape` has static length or not, which is what my modification does (`input_shape + (output_dim,)`)

The only difference here is when `input_shape[1]` is dynamic and `input_length` is static. But in this case, why would someone have a dynamic input?

**Comment by fchollet** [comment]
There are like a thousand use cases for having static-length sequences, starting with unrolled RNNs. This `input_length` parameter allows the model to do static shape inference.

**Comment by linxihui** [comment]
@fchollet but as I said above, whether the input_length should be static or not should depend on if the input_shape has static length, right?  Why would someone have a dynamic input for a static `input_length`?

**Comment by fchollet** [comment]
> whether the output length should be static or not should depend on if the input_shape[1] is static

That's true for the functional API. But the `Embedding` layer is often used as the first layer in a `Sequential` model. In this case, `input_length` would be the way to specify a static input length.

File: keras/layers/embeddings.py [hunk 1]
```diff
@@ -51,16 +51,16 @@ class Embedding(Layer):
           If mask_zero is set to True, as a consequence, index 0 cannot be
           used in the vocabulary (input_dim should equal size of
           vocabulary + 1).
-      input_length: Length of input sequences, when it is constant.
+      input_length: Lengths of input sequences, when it is constant.
```
Comment (fchollet) [inline keras/layers/embeddings.py]: If think it is confusing, especially for beginners (those who read the documentation) to have the docstring covered any other case than 2D. I would leave the docstring unchanged.

File: keras/layers/embeddings.py [hunk 2]
```diff
@@ -108,11 +108,25 @@ def compute_mask(self, inputs, mask=None):
...
     def compute_output_shape(self, input_shape):
-        if not self.input_length:
-            input_length = input_shape[1]
+        if self.input_length is None:
+            return input_shape + (self.output_dim,)
         else:
-            input_length = self.input_length
-        return (input_shape[0], input_length, self.output_dim)
+            # input_length can be tuple if input is 3D or higher
+            if hasattr(self.input_length, '__len__'):
```
Comment (fchollet) [inline keras/layers/embeddings.py]: `if isinstance(self.input_length, (tuple, list))`

**Comment by linxihui** [comment]
@fchollet  updated.

**Comment by fchollet** [comment]
LGTM, thanks

**Pull request merged**